### ISSUE TITLE
MAML implementation question

### ISSUE THREAD

**Issue by JeremyAlain** [deepchem/deepchem#1481]
@peastman 
I saw that you implemented the MAML code, thus I thought I would ask you as you are most familiar with it. 

I am not sure if I misunderstand something about your code, or if it is a mistake of this algorithm: 
<img width="530" alt="bildschirmfoto 2018-12-29 um 16 37 48" src="https://user-images.githubusercontent.com/20299322/50586638-b7905800-0e7a-11e9-83ec-44d1bf9c2a73.png">

That's how I understand the algorithm:
Assume our Neural Net is in a state X(i.e. the weights have certain values X). Then from this state X we train different tasks to get weight updates and based on these new weights we evaluate a new task. In line 10 we then take the sum of all these evaluations of specific tasks and calculate the gradient with respect to this total loss. But the important thing is that when we calculate the total gradient our Neural Net should again be in the state X.  

The way I understand your code, you are not doing this. Yes you do evaluate on the NeuralNet trained for every specific task. But then with these task-specific updated weights you again calculate a gradient. You then sum up all these gradients and apply them to the NeuralNet which is in State X. But the problem I see is that the gradients you calculate are not based on a NeuralNet in a state X but on different task specific optimized Neural Nets. 

Do I understand your implementation correclty? Is this on puropse? Am I right that the implementation is slightly different than in the paper? 

Cheers
Jérémy

**Comment by JeremyAlain** [comment]
EDIT: What I mean is specifically written here in the paper
<img width="537" alt="bildschirmfoto 2019-01-02 um 10 55 30" src="https://user-images.githubusercontent.com/20299322/50587134-ff17e380-0e7c-11e9-8730-3ef38d544044.png">

**Comment by JeremyAlain** [comment]
I have once more gone through the paper and have a suspiction: 
Did you implement the first order approximation of the MAML algorithm that they mention? Honestly I did not fully understand until now what that meant. But it seems to be what you implemented. Can you confirm that?

**Comment by peastman** [comment]
It's been a very long time since I worked on this, but I believe it's implementing the full algorithm, not the linear approximation.  The important point is that it creates two different sets of variables, one for the original parameters and one for the parameters after meta-optimization.  (See https://github.com/deepchem/deepchem/blob/master/deepchem/metalearning/maml.py#L143-L146.)  It computes the gradient of the meta-loss with respect to the *original* parameters.  Then it uses that to update the variables containing the original parameters.

**Comment by JeremyAlain** [comment]
Thanks for your answer, I looked through it again and you are right! 
kind regards
Jérémy
